File: src/javascript/algorithms/array/minAbsoluteDiff/minAbsoluteDiff.js
// Sort + scan adjacent — O(n log n) time, O(n) space
function minAbsDifference(nums) {
    nums.sort((a, b) => a - b);
    let minDiff = Infinity;
    for (let i = 1; i < nums.length; i++) {
        minDiff = Math.min(minDiff, nums[i] - nums[i - 1]);
    }
    const result = [];
    for (let i = 1; i < nums.length; i++) {
        if (nums[i] - nums[i - 1] === minDiff) {
            result.push([nums[i - 1], nums[i]]);
        }
    }
    return result;
}

const r1 = minAbsDifference([4, 2, 1, 3]);
const e1 = [[1,2],[2,3],[3,4]];
console.log(`Test 1: ${JSON.stringify(r1) === JSON.stringify(e1) ? '[PASS]' : '[FAIL]'} result=${JSON.stringify(r1)} expected=${JSON.stringify(e1)}`);

const r2 = minAbsDifference([1, 3, 6, 10, 15]);
const e2 = [[1,3]];
console.log(`Test 2: ${JSON.stringify(r2) === JSON.stringify(e2) ? '[PASS]' : '[FAIL]'} result=${JSON.stringify(r2)} expected=${JSON.stringify(e2)}`);

const r3 = minAbsDifference([3, 8, -10, 23, 19, -4, -14, 27]);
const e3 = [[-14,-10],[19,23],[23,27]];
console.log(`Test 3: ${JSON.stringify(r3) === JSON.stringify(e3) ? '[PASS]' : '[FAIL]'} result=${JSON.stringify(r3)} expected=${JSON.stringify(e3)}`);

const r4 = minAbsDifference([1, 2, 3]);
const e4 = [[1,2],[2,3]];
console.log(`Test 4: ${JSON.stringify(r4) === JSON.stringify(e4) ? '[PASS]' : '[FAIL]'} result=${JSON.stringify(r4)} expected=${JSON.stringify(e4)}`);

const r5 = minAbsDifference([-5, 0, 5]);
const e5 = [[-5,0],[0,5]];
console.log(`Test 5: ${JSON.stringify(r5) === JSON.stringify(e5) ? '[PASS]' : '[FAIL]'} result=${JSON.stringify(r5)} expected=${JSON.stringify(e5)}`);

File: src/python/algorithms/array/minAbsoluteDiff/min_absolute_diff.py
# Sort + scan adjacent — O(n log n) time, O(n) space
def min_abs_difference(nums):
    nums.sort()
    min_diff = float('inf')
    for i in range(1, len(nums)):
        min_diff = min(min_diff, nums[i] - nums[i - 1])
    result = []
    for i in range(1, len(nums)):
        if nums[i] - nums[i - 1] == min_diff:
            result.append([nums[i - 1], nums[i]])
    return result


if __name__ == "__main__":
    r1 = min_abs_difference([4, 2, 1, 3])
    e1 = [[1, 2], [2, 3], [3, 4]]
    print(f"Test 1: {'[PASS]' if r1 == e1 else '[FAIL]'} result={r1} expected={e1}")

    r2 = min_abs_difference([1, 3, 6, 10, 15])
    e2 = [[1, 3]]
    print(f"Test 2: {'[PASS]' if r2 == e2 else '[FAIL]'} result={r2} expected={e2}")

    r3 = min_abs_difference([3, 8, -10, 23, 19, -4, -14, 27])
    e3 = [[-14, -10], [19, 23], [23, 27]]
    print(f"Test 3: {'[PASS]' if r3 == e3 else '[FAIL]'} result={r3} expected={e3}")

    r4 = min_abs_difference([1, 2, 3])
    e4 = [[1, 2], [2, 3]]
    print(f"Test 4: {'[PASS]' if r4 == e4 else '[FAIL]'} result={r4} expected={e4}")

    r5 = min_abs_difference([-5, 0, 5])
    e5 = [[-5, 0], [0, 5]]
    print(f"Test 5: {'[PASS]' if r5 == e5 else '[FAIL]'} result={r5} expected={e5}")
